src/zkevm_specs/evm/execution/log.py

```diff
@@ -0,0 +1,67 @@
+from ..instruction import Instruction, Transition
+from ..table import CallContextFieldTag, TxLogFieldTag, TxContextFieldTag
+from ..opcode import Opcode
+from ..execution_state import ExecutionState
+from ...util.param import GAS_COST_LOG
+from ...util import FQ
+
+
+def log(instruction: Instruction):
+    opcode = instruction.opcode_lookup(True)
+    # constrain op in [log0, log4] range
+    instruction.range_lookup(opcode - Opcode.LOG0, 5)
+
+    # pop `mstart`, `msize` from stack
+    mstart = instruction.rlc_to_fq_exact(instruction.stack_pop(), 8)
+    msize = instruction.rlc_to_fq_exact(instruction.stack_pop(), 8)
+
+    # check contract_address in CallContext & TxLog
+    # use call context's  callee address as contract address
+    contract_address = instruction.call_context_lookup(CallContextFieldTag.CalleeAddress)
+    instruction.constrain_equal(contract_address, instruction.tx_log_lookup(TxLogFieldTag.Address))
+    # check not static call
+    instruction.constrain_equal(0, instruction.call_context_lookup(CallContextFieldTag.IsStatic))
+
```

Comment:
The only thing I find missing in this implementation is a read to CallContext -> IsPersistent, and then using that as a dynamic selector to disable (when is_persistent = 0):
1. All the lookups to TxLog rows in rw_table (and their associated increase in rw_counter)
2. The constraint that the next step must be `ExecutionState.CopyToLog`

Reply:
added in [a9b4e08](https://github.com/appliedzkp/zkevm-specs/pull/125/commits/a9b4e08f53226f01a0310de80b2366a6794129e9)

Reply:
Thanks!  I believe there's a more optimized way to handle the `is_persistent = False`: also skip the Memory reads as well as skipping all the `CopyToLog` steps.  Nevertheless I think the current implementation is correct so I'm OK with it!